scheduler/app/libs/covalent/treasury.py

```diff
@@ -55,6 +55,13 @@ async def get_treasury(portfolio: Dict[str, Any]) -> Treasury:
             )
         )
 
+    # These tokens don't receive a valid response from covalent's
+    # portfolio balances. So we blacklist them
+    blacklist = [
```

Comment:
I'd prefer avoiding hardcoding such things.

If I understand well, for these aseets, some of the `item['xxx']` below in the list comprehension are failing.

Also, we are anyways later in the flow removing any asset that is < 0.5%. Meaning that we don't care if small assets are failing.

So I'd recommend just doing something along the lines of:

```py
def build_erc20t(item):
    try:
        return ERC20(item['contract_name'],
            ...
    except KeyError:
        return None

assets = [
    erc20 for erc20 in [
        build_erc20(item) for item in portfolio["items"]
    ] if erc20 is not None
]
```

Would it suffice to solve the bug?

There is also a longer-term caveat here: if the Covalent API is failing for an asset that might be > 0.5% (or even just if the total of such failing assets is important in the treasury total) we might have wrong overall results (like the total of `treasury.assets` balances being noticeably inferior to the total balance). Let's just note this as a comment for now.

Reply:
> Also, we are anyways later in the flow removing any asset that is < 0.5%. Meaning that we don't care if small assets are failing.

Herein lies the issue. The `quote_rate` from the covalent response from those two blacklisted tokens gives the token amount in `1e18` erc20 decimals. The real issue actually cannot be corrected with what you suggested.

The response for this token `0x7cf56db0f7781d478d5a96f6ee8e0b5cbaaf8ad2` for example:

```json
{
"contract_decimals": 18,
"contract_name": "Omicron",
"contract_ticker_symbol":"OMIC",
"contract_address": "0x7cf56db0f7781d478d5a96f6ee8e0b5cbaaf8ad2",
"supports_erc": null,
"logo_url": "https://logos.covalenthq.com/tokens/1/0x7cf56db0f7781d478d5a96f6ee8e0b5cbaaf8ad2.png",
"holdings":
        [
            {
    "timestamp":"2022-06-07T00:00:00Z",
    "quote_rate":2.2089543E21,
    "open":     {
        "balance":"84945288674901668561641087", "quote":1.8764025E29},
    "high": "..."
}
```
[Here is the full response](https://api.covalenthq.com/v1/1/address/0xde21f729137c5af1b01d73af1dc21effa2b8a0d6/portfolio_v2/?&key=ckey_30f56650f3a544fe8803d522cb0). If you notice, the `quote` in the `holdings[0].open` property is 1.8764025**E29**. This value is supposed to be in USD but it is clearly in the ERC20 `decimals` format. I was attempting to filter out this error early in the control flow.

Letting this token stay in the portfolio will make all the relevant holdings be filtered out by your < 0.5% check. So your suggestion isn't sufficient.

This was something I wanted to discuss, how can we handle catching this exception better? Nothing stood out to me in the covalent response object body; `token_decimals` isn't a property on the response for example. `quote_currency` is however but it's clearly not the case in these two outliers.

Reply:
p.s. this does not fix the "blank chart" issue, but was found during investigating the original error.

Reply:
> Here is the full response. If you notice, the quote in the holdings[0].open property is 1.8764025E29. This value is supposed to be in USD but it is clearly in the ERC20 decimals format. I was attempting to filter out this error early in the control flow.

I understand better, thanks.

Since the underlying API is behaving pretty badly, let's keep the hardcoded values indeed, as I'm not sure how long we'll keep using it.

That said, I'm still wondering if we can do better: there are probably a lot of other bad-behaving tokens than the 2 that you've listed. Couldn't we implement a simple rule, like:

```py
if item['holdings'][0]['quote_rate'] > 10**18:
    # skip
```

Reply:
This is a better fix, I wanted to avoid assuming the number of decimals a token has since this can be set on token contract contstruction; a tokens decimals can be `1e6` as well e.g. USDC and other stable coins.

Turns out there is a `contract_decimals` property `for item in portfolio["items"]` so maybe this can work for more cases:

```python
if item['holdings'][0]['quote_rate'] > item["contract_decimals"]:
    # skip
```

Reply:
I actually wanted to avoid using this since a token contract's decimals can be different on contract creation; some tokens may have 12 decimals (1e12 or 10**12) and in this case the fix will not work.

However, your solution is suitable for the two misbehaving tokens in this case, and since 18 decimals is the ERC20 standard it will catch more cases than the initial fix.